File: 1. Learn the Basics/4. Hashing/C++/1. Frequency of elements in array.cpp
/*
Counting frequencies of array elements using hashing
*/

#include <bits/stdc++.h>
using namespace std;


int main() {
	// your code goes here
    int n;
    cin >> n;
    int arr[n];
    unordered_map<int,int> ump;
    map<int,int> mp;
    for(int i=0;i<n;i++){
        cin >> arr[i];
        ump[arr[i]]++;
        mp[arr[i]]++;
    }
    
    for(auto it : ump){
        cout << it.first << "->" << it.second << endl;
    }
    
    for(auto it : mp){
        cout << it.first << "->" << it.second << endl;
    }
    
    int q;
    cin >> q;
    while(q--){
        int number;
        cin >> number;
        cout << "Map " << number << " " << mp[arr[number]] <<endl;
        cout << "Unordered Map " << number << " " << ump[arr[number]] << endl;
    }
	return 0;
}

File: 1. Learn the Basics/4. Hashing/Java/1. Frequency of elements in array.java
/* package codechef; // don't place package name! */

import java.util.*;
import java.lang.*;
import java.io.*;

/* Name of the class has to be "Main" only if the class is public. */

/*
Counting frequencies of array elements using hashing
*/

class Codechef
{
    
	public static void main (String[] args) throws java.lang.Exception
	{
		// your code goes here
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int arr[] = new int[n];
		Map<Integer,Integer> mp = new HashMap<>();
		for(int i=0;i<n;i++){
		    arr[i] = sc.nextInt();
		    if(mp.containsKey(arr[i])){
		        mp.put(arr[i],mp.get(arr[i])+1);
		    } else {
		        mp.put(arr[i],1);
		    }
		}
	
		for(Map.Entry<Integer,Integer> entry : mp.entrySet()){
		    System.out.println(entry.getKey()+"->"+entry.getValue());
		}
		
		int q = sc.nextInt();
		while(q>0){
		    int number = sc.nextInt();
		    if(mp.containsKey(number)){
		        System.out.println(mp.get(number));
		    }else{
		        System.out.println(0);
		    }
		    q--;
		}
	}
}
